### PR TITLE
 ts-ify tonal-note

### PR DESCRIPTION
**Pull request by rakannimer**
Migrated tonal-note to typescript. 

Also migrated the tests to make sure the types work as expected. 

For tests where intentionally wrong arguments are passed I'm asking typescript to ignore them since we want the user to get a warning when they're mis-using the API. 

Please note that the type definitions are now generated.

Let me know what you think :) 

Cheers !

### CODE REVIEW

**Comment by danigb** [comment]
Wow! Great 🎉 Thanks! That's a lot of work! I've just took a quick look. Everything looks fantastic! There are some things that are wrong. For example:
```js
Note.pc("Bbbbbb4") // => Bbbbbb
```
I mean, pitch classes (pc function) and NoteNames are not limited to the values you assigned.

But the rest looks awesome! One question, the spec.js file is generated? Why two spec files?

Anyway, I'd ask you another thing: it seems that the CI is failing: https://travis-ci.org/danigb/tonal/builds/444683878

I don't know why. It looks like it can find the `tsc` compiler... can you try to solve it?

I'll review in depth as soon as I can...

Gracias!

**Comment by rakannimer** [comment]
Heyo ! 

Regarding pitch classes and note names I might be misunderstanding some things, I'm a self-taught musician so I'm learning as I go in the codebase :) 

I'll take a look and fix the tests and the pitch classes and NoteNames asap ! 

Cheers 😀

**Comment by danigb** [comment]
Hi @rakannimer!

This looks so good! 👍 🎉  We're almost there... I think types are now correct. But I'm a little bit worried about the infrastructure. Build tonal library is quite complex, and I would like to simplify the process instead of making even more complex.

Some questions:
- Is it possible to use ts-jest (or something similar) to avoid compiling the tests manually?
- Is it possible to use typescript compiler to create es5 and es6 versions of the code (so we can get rid of rollup?)

What are your thoughts about it?

**Comment by rakannimer** [comment]
Hey @danigb !

You're right the repo will become messier with 2 versions for each file.

> Is it possible to use ts-jest (or something similar) to avoid compiling the tests manually?

Yes, let me look into that.

> Is it possible to use typescript compiler to create es5 and es6 versions of the code (so we can get rid of rollup?)

Yes, we definitely can do that, but we'd need to change a bunch of build scripts in the packages' package.json and make each package in the mono-repo responsible for its own build process (right now we use the same build script for all packages). I'll setup a commit with a proposal we can go over. If you have a better idea, let me know !

Cheers !

**Comment by danigb** [comment]
Hi @rakannimer 

Thanks for taking care of this. 

> Migrating to ts-jest would break this implicit testing build pipeline.

I think we can assume we're going to use ts instead of babel. Or, if we go with the custom build process for each package, we can safely assume ts will be the language for the migrated packages (for both sources and tests.

> I think the best way to be able to make the build and testing less complex is by delegating the process to each package

The con I find is the repetition. But let's try this approach and improve it if we found a way to do it.

Thanks!

**Comment by danigb** [comment]
Hi @rakannimer 

I've been traveling these days. Now I'm back, and I'll take a look at your work this week. 👍

**Comment by rakannimer** [comment]
Hey @danigb !

Please take your time, I'm traveling too and will make the PR better the
longer you take :)

Cheers !

**Comment by danigb** [comment]
Wow! 👏 This looks really good! Ready to merge, isn't it? Let me take a close look just before.

File: packages/note/index.ts [hunk 1]
```diff
@@ -1,3 +1,37 @@
...
+type Note = string;
+type Midi = number;
+type Octave = number;
+type OrNull<T> = T | null;
+type NoteName = string;
+
+type NoteProps = {
+  name: NoteName;
+  letter: NoteLetter;
+  acc: NoteAccidental | string; // {String}: the note accidentals
```
Comment (danigb) [inline packages/note/index.ts]: Probably this is no longer required, isn't it?
```suggestion
  acc: NoteAccidental; // {String}: the note accidentals
```

**Comment by danigb** [comment]
Hi @rakannimer 

Happy to see this PR improving each time 👏 I made a last minute check and `yarn build` gave me this error:

```
index.ts → build/es6.js, build/es5.js...
rpt2: config error TS18003 No inputs were found in config file '/Users/dani/Code/Music/tonal/tsconfig.json'. Specified 'include' paths were '["./index.ts"]' and 'exclude' paths were '["./packages/**/*.d.ts","./packages/**/*.spec.ts"]'.
created build/es5.js in 107ms
```

But the build/* output files are actually created... any idea?

**Pull request merged**